Refactor the exec callback system

marcpayne opened the pull request:
### Overview

Prior to this commit, exec_cb callbacks were registered and the results were immediately fetched on every update interval... except the first. Besides not working on the first update, this meant that all exec_cb callbacks were re-registered on every update, which was not efficient.

The solution was to decouple the registration from the result retrieval and only register on the initial startup (or config reload). Now, the callbacks are registered when they are supposed to be, and the first-run issue is gone.

The code is hopefully more understandable now. I consolidated several functions in `exec.cc` into one, which makes the logic much easier to follow. I also added plenty of documentation, both to the code itself and to the manpage.

### Bonuses

  * execgraph works again!
  * optional execgraph parameters are now documented
  * execbar and execgauge now accept height and width parameters

### Known issues

  * technically, execgraph accepts commands with spaces, but such commands will need to be surrounded with double-quotes and you might as well forget adding optional arguments like colors, as the results may not be expected. Just put your spacey command into a script and run the script instead so that everything will just work.
  * the execgraph command comes *before* its arguments, whereas the execbar and execgauge commands come after their arguments.
  * texec{i} is registered on startup like everything else, but sometimes it will not produce output on the first interval. This is an intermittent issue.

### Example

![Exec Demo](https://www.dropbox.com/s/75gq1zodfjfif8u/conky-exec-demo.png?dl=1)

```
conky.text = [[
EXEC DEMO ${hr}
${exec date}

Default execbar w/ default_bar_height = 10
${execbar shuf -i 0-100 -n 1}

Variable size execbars
${execbar 20,150 shuf -i 0-90 -n 1}
${execbar 10,100 shuf -i 0-92 -n 1}
${execbar 3,50 shuf -i 0-94 -n 1}

Customized execgraphs
${execgraph ~/seconds.sh 25,150 FF0000 FFFF00 60 -t}${execgraph ~/random.sh 25,150 -l}

Variable size execgauges
${execgauge 50,80 shuf -i 0-100 -n 1}${execgauge 100,220 shuf -i 0-99 -n 1}
]];
```

marcpayne commented:
I already see changes that I should make, e.g. my enum would make more logical sense in exec.h.

Before I do that, I have a style question. Which do you prefer: `if (foo)` or `if(foo)`? Likewise with for/while. I like clearing up inconsistencies as I go.

KamranMackey commented:
@marcpayne I prefer `if (foo)` personally.

stardiviner commented:
I like it too.

marcpayne commented:
Yeah, I prefer that too. I added spaces in exec.cc where I did most of the work and left the rest as is. Also, I moved the enum over to exec.h today, so I think I'm done for now, barring any requests for changes.

KamranMackey commented:
@brndnmtthws this look good to you, Brenden?

brndnmtthws commented on src/text_object.h:
```diff
@@ -76,25 +77,37 @@ void gen_print_nothing(struct text_object *, char *, int);
 struct text_object {
 	struct text_object *next, *prev;	/* doubly linked list of text objects */
-	struct text_object *sub;		/* for objects parsing text into objects */
+	struct text_object *sub;			/* for objects parsing text into objects */
```
Looks like some whitespace misalignment here.

marcpayne replied:
Yeah, I noticed that the tabs don't align in the diffs. The reason I changed these lines is that they were misaligned in my editors (Kate and vim), which both use a tab width of 4 characters. Should I set the tab width to 8 instead? In any case, I'll fix the spacing so they line up here. Thanks!

brndnmtthws commented on src/text_object.h:
```diff
@@ -31,8 +31,9 @@
 #include <stdint.h>		/* uint8_t */
 #include "config.h"		/* for the defines */
-#include "specials.h"		/* enum special_types */
+#include "specials.h"	/* enum special_types */
```
Whitespace.

brndnmtthws commented on src/text_object.h:
```diff
@@ -76,25 +77,37 @@ void gen_print_nothing(struct text_object *, char *, int);
 
 class legacy_cb: public conky::callback<void *, int (*)()> {
-    typedef conky::callback<void *, int (*)()> Base;
+	typedef conky::callback<void *, int (*)()> Base;
```
Whitespace.

marcpayne replied:
Do you prefer hard tabs or soft tabs? Kate took the liberty of changing these tabs automatically, but I figured I would keep the changes, since most of the files seem to use hard tabs. Sorry for all the issues. I'm making note of style preferences for next time.

brndnmtthws replied:
Honestly, I don't care. Most of the existing codes uses hard tabs (i.e., `\t`), so we should stick with that or change all the tabs everywhere. If you want to do the latter, that should be in a separate PR.

brndnmtthws commented on src/update-cb.hh:
```diff
@@ -56,14 +56,15 @@ namespace conky {
-			const size_t hash;
-			uint32_t period;
-			uint32_t remaining;
+			const size_t hash;		// used to determined callback uniqueness
```
Whitespace, and `/* */` comment blocks are preferred.

marcpayne replied:
Ah, good to know about the `/* */` comment blocks! I've seen both, so I wasn't sure.

brndnmtthws commented:
Looks good, made a few stylistic suggestions.

marcpayne commented:
Okay, I cleaned up the whitespace issues and switched any new or modified comments to the `/* */` style. There are still some `//` comments around that I didn't touch, which I will leave for another PR someday.

I'll be happy to clear up any other issues if I missed something (not unusual). Thank you for your patience!

brndnmtthws commented:
Thank you.
